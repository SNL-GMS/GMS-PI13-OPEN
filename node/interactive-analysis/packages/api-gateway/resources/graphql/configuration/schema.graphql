
# Interface for analyst configuration used to determine fields based on roles
type AnalystConfiguration {
  logLevel: LogLevel!
  defaultNetwork: String!
  sohStationGroupNames: [SOHStationGroupNameWithPriority]!
  defaultFilters: [WaveformFilter]!
  redisplayPeriod: Int!
  reprocessingPeriod: Int!
  acknowledgementQuietDuration: Int!
  availableQuietDurations: [Int]!
  sohStationStaleTimeMS: Int!
  sohHistoricalDurations: [Int]!
  systemMessageLimit: Int!
}

# SOH StationGroup and Priority interface definition
type SOHStationGroupNameWithPriority {
  name: String
  priority: Int
}

# Represents a Waveform Filter
type WaveformFilter {
  id: String!
  name: String!
  description: String!
  filterType: String!
  filterPassBandType: String!
  lowFrequencyHz: Float
  highFrequencyHz: Float
  order: Float!
  filterSource: String!
  filterCausality: String!
  zeroPhase: String!
  sampleRate: Float!
  sampleRateTolerance: Float!
  validForSampleRate: String!
  aCoefficients: [Float]
  bCoefficients: [Float]
  groupDelaySecs: Float
}

# The following event related queries are supported.
extend type Query {
    uiAnalystConfiguration: AnalystConfiguration
}
