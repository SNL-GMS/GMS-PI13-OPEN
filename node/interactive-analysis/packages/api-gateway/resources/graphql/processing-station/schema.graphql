# Represents a group of stations used for monitoring.
# This is the processing equivalent of the ReferenceNetwork.
type ProcessingStationGroup {
    name: String
    description: String
    stations: [ProcessingStation]
}

# Represents an installation of monitoring sensors for the purposes of processing.
# Multiple sensors can be installed at the same station.
type ProcessingStation {
    name: String
    type: StationType
    description: String
    location: Location
    channelGroups: [ProcessingChannelGroup]
    channels: [ProcessingChannel]
}

# ChannelGroupType enum represents the different groupings of channels
enum ChannelGroupType {
    PROCESSING_GROUP
    SITE_GROUP
}

# Represents a physical installation (e.g., building, underground vault, borehole)
# containing a collection of Instruments that produce Raw Channel waveform data.
type ProcessingChannelGroup {
    name: String
    description: String
    location: Location
    type: ChannelGroupType
    channels: [ProcessingChannel]
}

# Represents a source for unprocessed (raw) or processed (derived) time series data
# from a seismic, hydroacoustic, or infrasonic sensor.
type ProcessingChannel {
    name: String
    displayName: String
    canonicalName: String
    description: String
    station: String
    channelDataType: ChannelDataType
    nominalSampleRateHz: Int
    location: Location
    orientationAngles: Orientation
}

# Enumeration representing the different types of processing channels.
enum ChannelDataType {
  SEISMIC
  HYDROACOUSTIC
  INFRASOUND
  WEATHER
  DIAGNOSTIC_SOH
  DIAGNOSTIC_WEATHER
}

# Represents the orientation angles used in processing channels
type Orientation {
    horizontalAngleDeg: Float
    verticalAngleDeg: Float
}

#####################################################
# Resolver entries Mutation, Query and Subscriptions
# The following channel-related queries are supported.
extend type Query {

    # Retrieve the default station list for use in automated and interactive analysis
    defaultProcessingStations: [ProcessingStation]
}


