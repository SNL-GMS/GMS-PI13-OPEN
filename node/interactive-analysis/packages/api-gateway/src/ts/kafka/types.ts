import { KeyValue } from '@gms/common-util';
import * as Immutable from 'immutable';

/**
 * Consumer Callback Function Type
 */
export type ConsumerCallback<T> = (topic: string, messages: Immutable.List<T>) => void;

/**
 * The KAFKA Status States
 */
export enum KafkaStatus {
  /** KAFKA has not been initialized */
  NOT_INITIALIZED = 'Not Initialized',
  /** KAFKA is connecting */
  CONNECTING = 'Connecting',
  /** KAFKA failed to connect and is reconnecting */
  RECONNECTING = 'Reconnecting',
  /** KAFKA has started */
  STARTED = 'Started',
  /** KAFKA is connected */
  CONNECTED = 'Connected',
  /** KAFKA has stopped */
  STOPPED = 'Stopped',
  /** KAFKA has disconnected */
  DISCONNECTED = 'Disconnected',
  /** KAFKA has as an error */
  ERROR = 'Error',
  /** KAFKA has crashed */
  CRASHED = 'Crashed'
}

/**
 * Defines the base interface for a KAFKA Consumer or Producer.
 */
// tslint:disable-next-line: interface-name
export interface IKafka {
  /**
   * Returns the KAFKA status.
   */
  getStatus(): KafkaStatus;

  /**
   * Returns the KAFKA status history information.
   */
  getStatusHistoryInformation(): Immutable.List<KeyValue<string, string | Error>>;

  /**
   * Returns the KAFKA status history information as an object
   */
  getStatusHistoryInformationAsObject(): {};

  /**
   * Returns the up time. The time in which KAFKA last established a good connection.
   */
  getUpTime(): Date;

  /**
   * Returns the number of seconds the KAFKA instance has been up and running (connected)
   */
  getUpTimeSeconds(): number;

  /**
   * Returns true if connected; false otherwise.
   */
  connected(): boolean;

  /**
   * Start the KAFKA consumer/producer.
   */
  start(): Promise<void>;

  /**
   * Stop the KAFKA consumer/producer.
   */
  stop(): Promise<void>;
}

/** Defines the KAFKA settings for the KAFKA Consumer */
export interface KafkaSettings {
  clientId: string;
  groupId: string;
  brokers: string[];
  connectionTimeout: number;
  maxWaitTimeInMs: number;
  heartbeatInterval: number;
  consumerTopics: {
    systemMessagesTopic: string;
    uiStationSoh: string;
  };
  producerTopics: {
    acknowledgedTopic: string;
    quietedTopic: string;
  };
}
