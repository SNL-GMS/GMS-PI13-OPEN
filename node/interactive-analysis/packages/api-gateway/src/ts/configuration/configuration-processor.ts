import { SOHStationGroupNameWithPriority } from '@gms/common-graphql/lib/graphql/ui-configuration/types';
import { getDurationMilliTime, sleep } from '@gms/common-util';
import { ApolloError } from 'apollo-server-core';
import config from 'config';
import Immutable from 'immutable';
import { gatewayLogger, gatewayLogger as logger } from '../log/gateway-logger';
import { HttpClientWrapper, HttpResponse } from '../util/http-wrapper';
import { WaveformFilterDefinition } from '../waveform-filter/model';
import * as configurationMockBackend from './configuration-mock-backend';
import {
  AnalystConfigs,
  AnalystConfiguration,
  CommonConfigs,
  CommonConfiguration,
  Configuration,
  defaultUserRole,
  Selector,
  SohConfiguration
} from './model';

/**
 * Singleton class used to retrieve Configuration attributes such as network and waveform filters
 * At start up a default network is fetched from the configuration service, if a user who logs in isn't the default
 * that users configuration will be fetched, at start up the station processors need a network and use a the default
 * network therefore the async method getDefaultNetwork is used, other wise after log in, can guarantee that
 * a users configuration will be loaded and the other get methods don't need to be awaited on.
 */
export class ConfigurationProcessor {
  /**
   * Returns the singleton instance of the cache processor.
   * @returns the instance of the cache processor
   */
  public static Instance(): ConfigurationProcessor {
    if (!ConfigurationProcessor.isInitialized) {
      ConfigurationProcessor.instance = new ConfigurationProcessor();
      ConfigurationProcessor.instance.initialize();
      ConfigurationProcessor.isInitialized = true;
    }
    return ConfigurationProcessor.instance;
  }

  /** The singleton instance */
  private static instance: ConfigurationProcessor;

  private static isInitialized: boolean = false;

  /**
   * The configuration
   */
  private readonly configuration: Configuration;

  /** Local configuration settings for reference stations */
  private readonly settings: any;

  /** HTTP client wrapper for communicating with backend services */
  private readonly httpWrapper: HttpClientWrapper;

  /**
   * Constructor - initialize the processor, loading settings and initializing the HTTP client wrapper.
   */
  private constructor() {
    // Load configuration settings
    this.settings = config.get('configuration');

    // Initialize an http client
    this.httpWrapper = new HttpClientWrapper();

    // Initialize the configuration
    this.configuration = {
      analystConfiguration: Immutable.Map(),
      sohConfiguration: undefined,
      commonConfiguration: Immutable.Map()
    };
  }

  /**
   * Gets the analyst configuration based on the role
   * @param role of the user from the user context
   * @returns a analyst configuration
   */
  public getAnalystConfiguration(role: string = defaultUserRole): AnalystConfiguration {
    if (this.configuration.analystConfiguration.has(role)) {
      return this.configuration.analystConfiguration.get(role);
    }
    throw new ApolloError(
      `Configuration error, could not find analyst configuration for role: ${role}`
    );
  }

  /**
   * Gets the common configuration based on the role
   * @param role of the user from the user context
   * @returns a common configuration
   */
  public getCommonConfiguration(role: string = defaultUserRole): CommonConfiguration {
    if (this.configuration.commonConfiguration.has(role)) {
      return this.configuration.commonConfiguration.get(role);
    }
    throw new ApolloError(
      `Configuration error, could not find common configuration for role: ${role}`
    );
  }

  /**
   * Gets the network based on the role
   * @param role of the user from the user context
   * @returns a network as a Promise<string>
   */
  public getNetworkForUser(role: string = defaultUserRole): string {
    if (this.configuration.analystConfiguration.has(role)) {
      return this.configuration.analystConfiguration.get(role).defaultNetwork;
    }
    throw new ApolloError(
      `Configuration error, could not find configuration default network for role: ${role}`
    );
  }

  /**
   * Gets the waveform filters based on the role
   * @param role of the user from the user context
   * @returns a waveform filters as Promise<WaveformFilterDefinition[]>
   */
  public getWaveformFiltersForUser(role: string = defaultUserRole): WaveformFilterDefinition[] {
    if (this.configuration.analystConfiguration.has(role)) {
      return this.configuration.analystConfiguration.get(role).defaultFilters;
    }
    throw new ApolloError(
      `Configuration error, could not find configuration default filters for role: ${role}`
    );
  }

  /**
   * Gets the soh configuration based on the role
   * @returns a soh configuration
   */
  public getSohConfiguration(): SohConfiguration {
    if (this.configuration.sohConfiguration) {
      return this.configuration.sohConfiguration;
    }
    throw new ApolloError(`Configuration error, could not find soh configuration`);
  }

  /**
   * Gets the soh configuration based on the role
   * @returns a soh configuration
   */
  public getStationGroupNamesWithPriorities(): SOHStationGroupNameWithPriority[] {
    if (!this.configuration.sohConfiguration) {
      return [];
    }
    return this.configuration.sohConfiguration.stationSohControlConfiguration.displayedStationGroups.map(
      (sg, count) => ({ name: sg, priority: count })
    );
  }

  /**
   * Gets the SOH Update Interval based on the role
   * @param role of the user from the user context
   * @returns the interval in seconds
   */
  public getSohUpdateIntervalForUser(role: string = defaultUserRole): number {
    if (this.configuration.sohConfiguration) {
      const sohParams = this.configuration.sohConfiguration.stationSohMonitoringDisplayParameters;
      return getDurationMilliTime(sohParams.redisplayPeriod);
    }
    throw new ApolloError(
      `Configuration error, could not find configuration redisplayPeriod for role: ${role}`
    );
  }

  /**
   * Gets the SOH quiet interval based on the role
   * @param role of the user from the user context
   * @returns the interval in seconds
   */
  public getSohDefaultQuietInterval(role: string = defaultUserRole): number {
    if (this.configuration.sohConfiguration) {
      const sohParams = this.configuration.sohConfiguration.stationSohMonitoringDisplayParameters;
      return getDurationMilliTime(sohParams.acknowledgementQuietDuration);
    }
    throw new ApolloError(
      `Configuration error, could not find configuration acknowledgementQuietDuration for role: ${role}`
    );
  }

  /**
   * Load and fetch the initial configurations for the given user role
   * @param role of the user from the user context
   */
  public async fetchConfiguration(): Promise<Configuration> {
    // if needed, fetch configurations
    await this.fetchAnalystConfiguration();
    await this.fetchSohConfiguration();
    await this.fetchCommonConfiguration();
    return this.configuration;
  }

  /**
   * Fetches AnalystConfiguration from the configuration service for the given unique configuration id
   * @param configuration the configuration id to retrieve
   */
  private async fetchAnalystConfiguration(
    role: string = defaultUserRole
  ): Promise<AnalystConfiguration> {
    if (this.configuration.analystConfiguration.has(role)) {
      return this.configuration.analystConfiguration.get(role);
    }

    // get the configuration id based on the user role
    const configuration: string = AnalystConfigs[role];

    logger.info(
      `Fetching analyst configuration data: role: ${role} configuration: ${configuration}`
    );
    const analystConfiguration = await this.internalFetchSubConfiguration<AnalystConfiguration>(
      configuration
    );
    this.configuration.analystConfiguration = this.configuration.analystConfiguration.set(
      role,
      analystConfiguration
    );
    return this.configuration.analystConfiguration.get(role);
  }

  /**
   * Fetches CommonConfiguration from the configuration service for the given unique configuration id
   * @param configuration the configuration id to retrieve
   */
  private async fetchCommonConfiguration(
    role: string = defaultUserRole
  ): Promise<CommonConfiguration> {
    if (this.configuration.commonConfiguration.has(role)) {
      return this.configuration.commonConfiguration.get(role);
    }

    // get the configuration id based on the user role
    const configuration: string = CommonConfigs[role];

    logger.info(
      `Fetching common configuration data: role: ${role} configuration: ${configuration}`
    );
    const commonConfiguration = await this.internalFetchSubConfiguration<CommonConfiguration>(
      configuration
    );
    this.configuration.commonConfiguration = this.configuration.commonConfiguration.set(
      role,
      commonConfiguration
    );
    return this.configuration.commonConfiguration.get(role);
  }

  /**
   * Fetches SohConfiguration from the configuration service for the given unique configuration id
   * @param configuration the configuration to retrieve
   */
  private async fetchSohConfiguration(): Promise<SohConfiguration> {
    // add data to the post because our service needs something in the body
    const postBodyData = 'test';
    // Retrieve the request configuration for the service call
    const requestConfiguration = this.settings.backend.services.getSohConfiguration.requestConfig;
    // Call the service and process the response data and add configuration to map
    const sohConfiguration = await this.callFetchSohConfiguration(
      requestConfiguration,
      postBodyData
    );
    this.configuration.sohConfiguration = sohConfiguration;
    return this.configuration.sohConfiguration;
  }

  /**
   * calls the webservice to get the sohconfiguration
   * @param requestConfiguration configuration given to the http request wrapper
   * @param postBodyData string to add to the boyd of the request,
   * the call needs a string in the body of the request
   */
  private readonly callFetchSohConfiguration = async (
    requestConfiguration: any,
    postBodyData: string
  ): Promise<SohConfiguration> =>
    this.httpWrapper
      .request<SohConfiguration>(requestConfiguration, postBodyData)
      .then((response: HttpResponse<SohConfiguration>) => response.data)
      .catch(async e => {
        logger.warn(`Error failed to connect to soh configuration service: ${e}`);
        const timeoutMS = 2000;
        return sleep<Promise<SohConfiguration>>(timeoutMS).then(async () => {
          logger.info(`Retrying to connect to configuration service`);
          logger.info(`retry config:: ${JSON.stringify(requestConfiguration)}`);
          return this.callFetchSohConfiguration(requestConfiguration, postBodyData);
        });
      })

  /**
   * Hits configuration service to get back an AnalystConfiguration based on role
   * @param configuration configuration to retrieve
   */
  private async internalFetchSubConfiguration<T>(
    configuration: string,
    selectors: Selector[] = []
  ): Promise<T> {
    // Build the query to be encoded as query string parameters
    const query = {
      configName: configuration,
      selectors
    };

    // Retrieve the request configuration for the service call
    const requestConfig = this.settings.backend.services.getAnalystConfiguration.requestConfig;

    logger.debug(
      `Requesting Analyst configuration query ${JSON.stringify(query)} config ${JSON.stringify(
        requestConfig
      )}`
    );

    // Call the service and process the response data and add configuration to map
    return this.httpWrapper
      .request<T>(requestConfig, query)
      .then((response: HttpResponse<T>) => response.data)
      .catch(async e => {
        logger.warn(`Error failed to connect to configuration service: ${e}`);
        const timeoutMS = 2000;

        return sleep<Promise<AnalystConfiguration>>(timeoutMS).then(async () => {
          logger.info(`Retrying to connect to configuration service`);
          return this.internalFetchSubConfiguration<T>(configuration, selectors);
        });
      });
  }

  /**
   * Initialize the configuration processor, fetching configuration from the backend.
   * This function sets up a mock backend if configured to do so.
   */
  private initialize() {
    try {
      logger.info(
        'Initializing the configuration processor - Mock Enable: %s',
        this.settings.backend.mock.enable
      );

      // If service mocking is enabled, initialize the mock backend
      if (this.settings.backend.mock.enable) {
        configurationMockBackend.initialize(this.httpWrapper.createHttpMockWrapper());
      }
    } catch (error) {
      gatewayLogger.error(`Failed to initialize the Configuration processor: ${error}`);
      throw error;
    }
  }
}
