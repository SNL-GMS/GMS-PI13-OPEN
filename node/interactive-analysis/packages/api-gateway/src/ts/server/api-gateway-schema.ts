import { makeExecutableSchema } from 'apollo-server-express';
import merge from 'lodash/merge';
import path from 'path';
import { resolvers as cacheResolvers } from '../cache/resolvers';
import { schema as cacheSchema } from '../cache/schema';
import { resolvers as channelSegmentResolvers } from '../channel-segment/resolvers';
import { schema as channelSegmentSchema } from '../channel-segment/schema';
import {
  extendedResolvers as commonExtendedResolvers,
  resolvers as commonResolvers
} from '../common/resolvers';
import { extendedSchema as commonExtendedSchema, schema as commonSchema } from '../common/schema';
import { resolvers as configurationResolvers } from '../configuration/resolvers';
import { schema as configurationSchema } from '../configuration/schema';
import { resolvers as dataAcquisitionStatusResolvers } from '../data-acquisition/resolvers';
import { schema as dataAcquisitionStatusSchema } from '../data-acquisition/schema';
import { schema as eventSchema } from '../event/model-and-schema/schema';
import { resolvers as eventResolvers } from '../event/resolvers';
import { resolvers as fkResolvers } from '../fk/resolvers';
import { schema as fkSchema } from '../fk/schema';
import { resolvers as qcMaskResolvers } from '../qc-mask/resolvers';
import { schema as qcMaskSchema } from '../qc-mask/schema';
import { resolvers as signalDetectionResolvers } from '../signal-detection/resolvers';
import { schema as signalDetectionSchema } from '../signal-detection/schema';
import { resolvers as sohStatusResolvers } from '../soh/resolvers';
import { schema as sohStatusSchema } from '../soh/schema';
import { resolvers as processingStationResolvers } from '../station/processing-station/resolvers';
import { schema as processingStationSchema } from '../station/processing-station/schema';
import { resolvers as referenceStationResolvers } from '../station/reference-station/resolvers';
import { schema as referenceStationSchema } from '../station/reference-station/schema';
import { resolvers as systemMessageResolvers } from '../system-message/resolvers';
import { schema as systemMessageSchema } from '../system-message/schema';
import { resolvers as userProfileResolvers } from '../user-profile/resolvers';
import { schema as userProfileSchema } from '../user-profile/schema';
import { resolvers as waveformResolvers } from '../waveform/resolvers';
import { schema as waveformSchema } from '../waveform/schema';
import { resolvers as workflowResolvers } from '../workflow/resolvers';
import { schema as workflowSchema } from '../workflow/schema';

/**
 * Global GraphQL schema definition for the entire API gateway
 */

// using the gateway logger
import { gatewayLogger as logger } from '../log/gateway-logger';
const objectPath = path.relative(process.cwd(), __filename);

// GraphQL schema definitions
logger.info('Creating graphql schema...', { module: objectPath });

const typeDefs = [
  commonSchema,
  commonExtendedSchema,
  systemMessageSchema,
  workflowSchema,
  processingStationSchema,
  referenceStationSchema,
  waveformSchema,
  signalDetectionSchema,
  eventSchema,
  qcMaskSchema,
  userProfileSchema,
  channelSegmentSchema,
  fkSchema,
  dataAcquisitionStatusSchema,
  sohStatusSchema,
  cacheSchema,
  configurationSchema
];

// Merge GraphQL resolvers from the schema domains
const resolvers = merge(
  commonResolvers,
  commonExtendedResolvers,
  systemMessageResolvers,
  workflowResolvers,
  processingStationResolvers,
  referenceStationResolvers,
  waveformResolvers,
  signalDetectionResolvers,
  eventResolvers,
  userProfileResolvers,
  qcMaskResolvers,
  channelSegmentResolvers,
  fkResolvers,
  dataAcquisitionStatusResolvers,
  sohStatusResolvers,
  cacheResolvers,
  configurationResolvers
);

// Build the GraphQL schema
export const schema = makeExecutableSchema({
  typeDefs,
  resolvers
});
