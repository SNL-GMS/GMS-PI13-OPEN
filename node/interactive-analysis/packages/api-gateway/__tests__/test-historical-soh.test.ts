import { SohMonitorType } from '@gms/common-graphql/lib/graphql/soh/types';
import { graphql } from 'graphql';
import { ConfigurationProcessor } from '../src/ts/configuration/configuration-processor';
import { schema } from '../src/ts/server/api-gateway-schema';
import { SohProcessor } from '../src/ts/soh/soh-processor';
import { ProcessingStationProcessor } from '../src/ts/station/processing-station/processing-station-processor';
import { userContext } from './__data__/user-profile-data';

// tslint:disable-next-line: no-magic-numbers
Date.now = jest.fn().mockReturnValue(1575410988600);

/**
 * Sets up test by loading SOH data
 */
beforeEach(async () => setupTest());
async function setupTest() {
  await ConfigurationProcessor.Instance().fetchConfiguration();
  await ProcessingStationProcessor.Instance().fetchStationData();
  SohProcessor.Instance();
}

const buildQuery = (types: SohMonitorType[]): string =>
  `query historicalSohByStationQuery {
    historicalSohByStation(queryInput:
      {
        stationName: "AAK",
        startTime: 1274392801,
        endTime: 1274400000,
        sohMonitorTypes: [${String(types)}]
      } ) {
      stationName
      monitorValues {
        channelName
        valuesByType {
          LAG {
            type
          }
          MISSING {
            type
          }
        }
      }
    }
  }
  `;

describe('Fetch SOH Historical data', () => {
  it('loads station soh historical soh LAG and MISSING data', async () => {
    // Execute the GraphQL query
    const query = buildQuery([SohMonitorType.LAG, SohMonitorType.MISSING]);
    const rootValue = {};
    const result = await graphql(schema, query, rootValue, userContext);
    const { data } = result;
    expect(data).toMatchSnapshot();
  });

  it('loads station soh historical soh only LAG data', async () => {
    // Execute the GraphQL query
    const query = buildQuery([SohMonitorType.LAG]);
    const rootValue = {};
    const result = await graphql(schema, query, rootValue, userContext);
    const { data } = result;
    expect(data).toMatchSnapshot();
  });

  it('loads station soh historical soh only MISSING data', async () => {
    // Execute the GraphQL query
    const query = buildQuery([SohMonitorType.MISSING]);
    const rootValue = {};
    const result = await graphql(schema, query, rootValue, userContext);
    const { data } = result;
    expect(data).toMatchSnapshot();
  });

  it('loads station soh historical NO soh data', async () => {
    // Execute the GraphQL query using an env entry
    const query = buildQuery([SohMonitorType.ENV_CLIPPED]);
    const rootValue = {};
    const result = await graphql(schema, query, rootValue, userContext);
    const { data } = result;
    expect(data).toMatchSnapshot();
  });
});
