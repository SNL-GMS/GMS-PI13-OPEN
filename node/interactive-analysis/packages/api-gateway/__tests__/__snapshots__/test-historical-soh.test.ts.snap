// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fetch SOH Historical data loads station soh historical NO soh data 1`] = `
Object {
  "historicalSohByStation": Object {
    "monitorValues": Array [
      Object {
        "channelName": "AAK.AAK.BHN",
        "valuesByType": Object {
          "LAG": null,
          "MISSING": null,
        },
      },
      Object {
        "channelName": "AAK.AAK.BHZ",
        "valuesByType": Object {
          "LAG": null,
          "MISSING": null,
        },
      },
      Object {
        "channelName": "AAK.AAK.BHE",
        "valuesByType": Object {
          "LAG": null,
          "MISSING": null,
        },
      },
    ],
    "stationName": "AAK",
  },
}
`;

exports[`Fetch SOH Historical data loads station soh historical soh LAG and MISSING data 1`] = `
Object {
  "historicalSohByStation": Object {
    "monitorValues": Array [
      Object {
        "channelName": "AAK.AAK.BHN",
        "valuesByType": Object {
          "LAG": Object {
            "type": "DURATION",
          },
          "MISSING": Object {
            "type": "PERCENT",
          },
        },
      },
      Object {
        "channelName": "AAK.AAK.BHZ",
        "valuesByType": Object {
          "LAG": Object {
            "type": "DURATION",
          },
          "MISSING": Object {
            "type": "PERCENT",
          },
        },
      },
      Object {
        "channelName": "AAK.AAK.BHE",
        "valuesByType": Object {
          "LAG": Object {
            "type": "DURATION",
          },
          "MISSING": Object {
            "type": "PERCENT",
          },
        },
      },
    ],
    "stationName": "AAK",
  },
}
`;

exports[`Fetch SOH Historical data loads station soh historical soh only LAG data 1`] = `
Object {
  "historicalSohByStation": Object {
    "monitorValues": Array [
      Object {
        "channelName": "AAK.AAK.BHN",
        "valuesByType": Object {
          "LAG": Object {
            "type": "DURATION",
          },
          "MISSING": null,
        },
      },
      Object {
        "channelName": "AAK.AAK.BHZ",
        "valuesByType": Object {
          "LAG": Object {
            "type": "DURATION",
          },
          "MISSING": null,
        },
      },
      Object {
        "channelName": "AAK.AAK.BHE",
        "valuesByType": Object {
          "LAG": Object {
            "type": "DURATION",
          },
          "MISSING": null,
        },
      },
    ],
    "stationName": "AAK",
  },
}
`;

exports[`Fetch SOH Historical data loads station soh historical soh only MISSING data 1`] = `
Object {
  "historicalSohByStation": Object {
    "monitorValues": Array [
      Object {
        "channelName": "AAK.AAK.BHN",
        "valuesByType": Object {
          "LAG": null,
          "MISSING": Object {
            "type": "PERCENT",
          },
        },
      },
      Object {
        "channelName": "AAK.AAK.BHZ",
        "valuesByType": Object {
          "LAG": null,
          "MISSING": Object {
            "type": "PERCENT",
          },
        },
      },
      Object {
        "channelName": "AAK.AAK.BHE",
        "valuesByType": Object {
          "LAG": null,
          "MISSING": Object {
            "type": "PERCENT",
          },
        },
      },
    ],
    "stationName": "AAK",
  },
}
`;
